crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -107,8 +107,65 @@ def cancel
         }
       end
 
+      # Orchestrate the upgrade of the nodes
+      def nodes
+        # check for current global status
+        # 1. TODO: return if upgrade has finished
+        # 2. TODO: find the next big step
+        next_step = "controllers"
+
+        if next_step == "controllers"
+
+          # TODO: Save the "current_step" to global status
+          if upgrade_controller_nodes
+            # upgrading controller nodes succeeded, we can continue with computes
+            next_step = "computes"
+          else
+            # upgrading controller nodes has failed, exiting
+            # leaving next_step as "controllers", so we continue from correct point on retry
+            return false
+          end
+        end
+
+        if next_step == "computes"
+          # TODO: Save the "current_step" to global status
+          upgrade_compute_nodes
+        end
+        true
+      end
+
       protected
 
+      def upgrade_controller_nodes
+        # TODO: find the controller node that needs to be upgraded now
+        # First node to upgrade is DRBD slave
+        drbd_slave = ""
+        NodeObject.find(
+          "state:crowbar_upgrade AND (roles:database-server OR roles:rabbitmq-server)"
+        ).each do |db_node|
+          cmd = "crm resource status ms-drbd-{postgresql,rabbitmq}| grep \\$(hostname) | grep -q Master"
```

Comment:
And I'd rather you didn't :-)  There's really no good reason not to split the line.  Just before the `| grep` would be the nicest place to do it.